text/0000-impl-trait-type-aliases.md

```diff
@@ -0,0 +1,380 @@
+
+## Type alias
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+```
+
+`impl Trait` type aliases are useful for declaring types that are constrained by traits, but whose concrete type should be a hidden implementation detail. We can use it in place of return-position `impl Trait` as in the previous examples.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+// The same as `fn bar() -> impl Baz`
+fn bar() -> Baz {
+    // ...
+}
+```
+
+However, if we use `Baz` in multiple locations, we constrain the concrete type referred to by `Baz` to be the same, so we get a type that we know will be the same everywhere and will satisfy specific bounds, whose concrete type is hidden. This can be useful in libraries where you want to hide implementation details.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+impl Trait for u8 {}
+
+fn foo() -> Baz {
+    let x: u8;
+    // ...
+    x
+}
+
+fn bar(x: Baz, y: Baz) {
+    // ...
+}
+
+struct Foo {
+    a: Baz,
+    b: (Baz, Baz),
+}
+```
+
+In this example, the concrete type referred to by `Baz` is guaranteed to be the same wherever `Baz` occurs.
+
+Note that using `Baz` as an argument type is *not* the same as argument-position `impl Trait`, as `Baz` refers to a unique type, whereas the concrete type for argument-position `impl Trait` is determined by the caller.
+
+```rust
+trait Trait {}
+
+type Baz = impl Trait;
+
+fn foo(x: Baz) {
+    // ...
+}
+
+// is *not* the same as:
+
+fn foo(x: impl Trait) {
+    // ...
+}
+```
+
+Just like with any other type alias, we can use `impl Trait` to specify associated types for traits, as in the following example.
+
+```rust
+trait Trait {
+    type Assoc;
+}
+
+struct Foo {}
+
+impl Trait for Foo {
+    type Assoc = impl Debug;
+}
+```
+
+Here, anything that makes use of `Foo` knows that `Foo::Assoc` implements `Debug`, but has no knowledge of its concrete type.
+
+[Eventually, we would also describe the use of `impl Trait` in `let`, `const` and `static` bindings, but as they are as-yet unimplemented and function the same as return-type `impl Trait`, they haven't been included here.]
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Since RFC 2071 was accepted, the initial implementation of `existential type` [has already been completed](https://github.com/rust-lang/rust/pull/52024). This RFC would simply replace the syntax of `existential type`, from:
+
+```rust
+existential type Foo: Bar;
+```
+
+to:
+
+```rust
+type Foo = impl Bar;
+```
+
+In addition, when documenting `impl Trait`, explanations of the feature would avoid type theoretic terminology (specifically "existential types") and prefer type inference language (if any technical description is needed at all).
+
+## Restricting compound `impl Trait` trait aliases
```

Comment:
This feels like somewhat an arbitrary restriction. Extending this to work anywhere in the definition should be fairly easy to do by recursively expanding each occurrence of `impl Trait` to a new fresh alias (+ generics) until there are only base cases left. I get being conservative, but not for it's own sake.

I would like to see an unresolved question about this at least which we may deal with during implementation.

Reply:
Ah, it would be better as an unresolved question; you're right.